File: 2018/2/2_1.py
def letter_counts(line):
    counts = {}
    for value in line:
        counts[value] = counts.setdefault(value, 0) + 1
    return counts


with open('./input') as lines:
    count_2 = 0
    count_3 = 0
    for line in lines:
        counts = letter_counts(line)
        if [v for v in counts.values() if v == 2]:
            count_2 += 1
        if [v for v in counts.values() if v == 3]:
            count_3 += 1

    print(count_2 * count_3)

File: 2018/2/2_1.clj
(use 'clojure.java.io)

(defn letter_counts [line]
  (reduce
   (fn [counts letter]
     (update counts letter (fn [c] (+ (if (nil? c) 0 c) 1))))
   {} line))

(defn has? [line_counts n]
  (contains? (set (vals line_counts)) n))

(defn counts [line_counts]
  {:2 (if (has? line_counts 2) 1 0)
   :3 (if (has? line_counts 3) 1 0)})

(def total (partial
            reduce
            (fn [c1 c2] {:2 (+ (get c1 :2)
                               (get c2 :2))
                         :3 (+ (get c1 :3)
                               (get c2 :3))})))

(with-open [r (reader "./input")]
  (->> r
       line-seq
       (map letter_counts)
       (map counts)
       total
       ((fn [total] (* (get total :2)
                       (get total :3))))
       println))
